### PR TITLE
Add deterministic RL training experiment with vLLM

### PR DESCRIPTION
**Pull request by bwasti**
This experiment provides a complete framework for bitwise-deterministic reinforcement learning training that combines vLLM for fast rollouts and TorchTitan for training with gradients.

Key features:
- Bitwise-deterministic forward and backward passes
- vLLM-compatible Qwen3 model with merged projections
- Custom Flash Attention with gradient support
- Gradient support for vLLM's batch-invariant operations
- Complete RL training loop with GRPO-style advantages
- Comprehensive test suite for determinism verification

Components:
- models/attention.py: VLLMCompatibleFlashAttention
- models/qwen3/model_vllm_compat.py: vLLM-compatible Qwen3 model
- batch_invariant_backward.py: Gradient support for vLLM operations
- simple_rl.py: End-to-end RL training loop
- tests/: Test suite for backward passes and determinism

### CODE REVIEW

File: torchtitan/experiments/deterministic_vllm_rl/README.md [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+2. Compute rewards based on the samples
+3. Update the policy using gradients
+
+**The problem**: If the forward pass during training differs from the forward pass during rollout, the gradients will be wrong! This is especially important for PPO, which compares old and new policy probabilities.
+
+**The solution**: Use the same deterministic kernels for both rollouts (vLLM) and training (TorchTitan). This ensures that `logprobs_rollout == logprobs_training` (bitwise).
+
+### Performance
+
+- **Rollout speed**: ~100x faster than standard PyTorch (thanks to vLLM)
```
Comment (tianyu-l) [inline torchtitan/experiments/deterministic_vllm_rl/README.md]: 👍  lol

File: torchtitan/experiments/deterministic_vllm_rl/batch_invariant_backward.py
Comment (tianyu-l) [inline torchtitan/experiments/deterministic_vllm_rl/batch_invariant_backward.py:1]: could you help add the disclaimer to this file too?
```
# Copyright (c) Meta Platforms, Inc. and affiliates.
# All rights reserved.
#
# This source code is licensed under the BSD-style license found in the
# LICENSE file in the root directory of this source tree.
```

File: torchtitan/experiments/deterministic_vllm_rl/simple_rl.py
Comment (tianyu-l) [inline torchtitan/experiments/deterministic_vllm_rl/simple_rl.py:1]: similar, and also the two test files
```
# Copyright (c) Meta Platforms, Inc. and affiliates.
# All rights reserved.
#
# This source code is licensed under the BSD-style license found in the
# LICENSE file in the root directory of this source tree.
```

File: torchtitan/experiments/deterministic_vllm_rl/README.md
Comment (tianyu-l) [inline torchtitan/experiments/deterministic_vllm_rl/README.md:1]: Can we mention that this only works for single device right now, and we plan to extend it to work with parallelisms?

File: torchtitan/experiments/deterministic_vllm_rl/batch_invariant_backward.py [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+        # Save for backward
+        ctx.save_for_backward(input, weight)
+        ctx.eps = eps
+
+        return output
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        """
+        Backward pass using batch-invariant PyTorch operations.
```
Comment (tianyu-l) [inline torchtitan/experiments/deterministic_vllm_rl/batch_invariant_backward.py:147]: Batch-invariant backward sounds nice! But I guess batch-invariance doesn't matter too much for backward? E.g. you'll lose batch invariance when doing loss reduction?

Asking because the decomposed implementation below might be slower than a non-batch-invariant kernel.

File: torchtitan/experiments/deterministic_vllm_rl/batch_invariant_backward.py [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+            grad_weight = torch.matmul(grad_output.transpose(0, 1), input)
+
+    if compute_grad_bias:
+        # grad_bias = sum(grad_output) along all dims except last
+        grad_bias = grad_output.sum(dim=tuple(range(grad_output.ndim - 1)))
+
+    return grad_input, grad_weight, grad_bias
+
+
+def rms_norm_backward_impl(grad_output, input, weight, eps):
```
Comment (tianyu-l) [inline torchtitan/experiments/deterministic_vllm_rl/batch_invariant_backward.py]: Is this used, as I didn't see it registered to backward op?

File: torchtitan/experiments/deterministic_vllm_rl/batch_invariant_backward.py [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+            False,            # deterministic
+            True,             # rotary_interleaved
+            ctx.scheduler_metadata,
+        )
+
+        # Return gradients for all forward inputs (None for non-tensor args)
+        return grad_q, grad_k, grad_v, None, None, None, None, None, None, None, None, None, None, None
+
+
+class RMSNormFunction(Function):
```
Comment (wwwjn) [inline torchtitan/experiments/deterministic_vllm_rl/batch_invariant_backward.py:114]: This function has been included in qwen3/model_vllm_compat/py, what's the differences between these 2?

File: torchtitan/experiments/deterministic_vllm_rl/batch_invariant_backward.py [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+        """
+        q, k, v, out, softmax_lse, cu_seqlens_q, cu_seqlens_k = ctx.saved_tensors
+
+        # Allocate gradient tensors
+        grad_q = torch.empty_like(q)
+        grad_k = torch.empty_like(k)
+        grad_v = torch.empty_like(v)
+
+        # Call FA3 backward kernel
+        torch.ops._vllm_fa3_C.bwd(
```
Comment (tianyu-l) [inline torchtitan/experiments/deterministic_vllm_rl/batch_invariant_backward.py]: curious: can this not be done also via
```
_batch_invariant_backward_LIB.impl("aten::matmul_backward", matmul_backward_impl, "CUDA")
```
but registering / overriding the vllm fa3 bwd op?

File: torchtitan/experiments/deterministic_vllm_rl/models/attention.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+        # Create cumulative sequence lengths
+        # cu_seqlens: [0, seq_len, 2*seq_len, ..., batch_size*seq_len]
+        cu_seqlens = torch.arange(
+            0, (batch_size + 1) * seq_len, seq_len,
+            dtype=torch.int32, device=q.device
+        )
+
+        # Wrap Flash Attention with manual backward pass
+        class FlashAttnWithBackward(torch.autograd.Function):
```
Comment (tianyu-l) [inline torchtitan/experiments/deterministic_vllm_rl/models/attention.py:60]: What's the relationship between the two `autograd.Function`: `FlashAttn3Function` and `FlashAttnWithBackward`. My naive feeling was that we only need one of them -- we either hand-write the entire backward (`FlashAttnWithBackward`), or we register a backward kernel for the forward kernel (FlashAttn3Function).

File: torchtitan/experiments/deterministic_vllm_rl/simple_rl.py [hunk 7]
```diff
@@ -0,0 +1,883 @@
...
+import tempfile
+import torch
+import torch.nn.functional as F
+from transformers import AutoTokenizer, AutoConfig
+from safetensors.torch import load_file, save_file
+from huggingface_hub import snapshot_download
+import numpy as np
+from torch.utils.tensorboard import SummaryWriter
+
+import torchtitan.experiments.compat
```
Comment (tianyu-l) [inline torchtitan/experiments/deterministic_vllm_rl/simple_rl.py]: not used and will error?

File: torchtitan/experiments/deterministic_vllm_rl/simple_rl.py [hunk 8]
```diff
@@ -0,0 +1,883 @@
...
+from torchtitan.models.qwen3.model.args import Qwen3ModelArgs
+from weights_vllm_compat import torchtitan_to_vllm_compat, vllm_compat_to_torchtitan
```
Comment (tianyu-l) [inline torchtitan/experiments/deterministic_vllm_rl/simple_rl.py]: These files are missing?

File: torchtitan/experiments/deterministic_vllm_rl/simple_rl.py [hunk 9]
```diff
@@ -0,0 +1,883 @@
...
+from weights.converter import torchtitan_to_vllm, vllm_to_torchtitan
```
Comment (tianyu-l) [inline torchtitan/experiments/deterministic_vllm_rl/simple_rl.py]: same

File: torchtitan/experiments/deterministic_vllm_rl/simple_rl.py [hunk 10]
```diff
@@ -0,0 +1,883 @@
...
+    from weights_vllm_compat import torchtitan_to_vllm_compat
+    titan_state = model.state_dict()
+    vllm_compat_state = torchtitan_to_vllm_compat(titan_state)
+    vllm_engine.update_weights(vllm_compat_state)
+
+    # Round-trip: load weights back from disk to maintain consistency with vLLM
+    import glob
+    from safetensors.torch import load_file as sf_load
+    from weights.converter import vllm_to_torchtitan
+    from weights_vllm_compat import torchtitan_to_vllm_compat as titan_to_vllm_compat
```
Comment (tianyu-l) [inline torchtitan/experiments/deterministic_vllm_rl/simple_rl.py]: this seems the same as few lines above

File: torchtitan/experiments/deterministic_vllm_rl/batch_invariant_backward.py (hunk 5 above)
Comment (tianyu-l) [reply torchtitan/experiments/deterministic_vllm_rl/batch_invariant_backward.py]: also curious that since the op itself already has a backward pass, why do we need to explicitly register an autograd function with both forward op and backward op?

File: torchtitan/experiments/deterministic_vllm_rl/simple_rl.py [hunk 11]
```diff
@@ -0,0 +1,883 @@
...
+
+        if use_vllm_compat:
+            # Convert to vLLM-compat format for vLLM-compatible model
+            weights_for_model = titan_to_vllm_compat(titan_from_disk)
+        else:
+            # Use standard TorchTitan format for standard model
+            weights_for_model = titan_from_disk
+
+        # Load back into model
+        model.load_state_dict(weights_for_model, strict=False)
```
Comment (tianyu-l) [inline torchtitan/experiments/deterministic_vllm_rl/simple_rl.py]: This looks a bit confusing. I thought loading from disk should be done only once, instead of in each `rl_update_step`. Is it related to the checkpoint bug you mentioned earlier?

File: torchtitan/experiments/deterministic_vllm_rl/batch_invariant_backward.py [hunk 12]
```diff
@@ -0,0 +1,473 @@
...
+        else:
+            # Fall through to original implementation for FA2
+            return _original_flash_attn_varlen_func(*args, **kwargs)
+
+    fa_interface.flash_attn_varlen_func = patched_flash_attn_varlen_func
+
+    _batch_invariant_backward_MODE = True
+
+
+def enable_batch_invariant_backward_mode():
```
Comment (wwwjn) [inline torchtitan/experiments/deterministic_vllm_rl/batch_invariant_backward.py]: Why do we need this renaming?

File: torchtitan/experiments/deterministic_vllm_rl/batch_invariant_backward.py (hunk 4 above)
Comment (teja-rao) [reply torchtitan/experiments/deterministic_vllm_rl/batch_invariant_backward.py:114]: these are same i think, one of them should be cleaned up.

File: torchtitan/experiments/deterministic_vllm_rl/simple_rl.py (hunk 11 above)
Comment (teja-rao) [reply torchtitan/experiments/deterministic_vllm_rl/simple_rl.py]: so I believe this is done to because we suspected that there may be some rounding issues in weights transfer but i verified removing this code and it works fine. my experience on checkpointing, unless there is a dtype conversion, weight loading should always be bitwise in sync. so, i will send an update removing this.

File: torchtitan/experiments/deterministic_vllm_rl/batch_invariant_backward.py (hunk 4 above)
Comment (bwasti) [reply torchtitan/experiments/deterministic_vllm_rl/batch_invariant_backward.py:114]: sorry that was a rebase conflict

File: torchtitan/experiments/deterministic_vllm_rl/README.md (hunk 1 above)
Comment (bwasti) [reply torchtitan/experiments/deterministic_vllm_rl/README.md]: this is a totally random number lmao -- kv cache does all the work.  thanks claude for adding this

File: torchtitan/experiments/deterministic_vllm_rl/README.md [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+- Training speed: Similar to standard TorchTitan
+- Memory: Saves activations for custom backward passes
+
+### Limitations
+
+1. Custom backward requires uniform sequence lengths
+2. Only causal attention is supported
+3. Requires NVIDIA GPUs with Flash Attention support
+
+## Project Structure
```
Comment (tianyu-l) [inline torchtitan/experiments/deterministic_vllm_rl/README.md:214]: It seems this chart needs to be updated.

File: torchtitan/experiments/deterministic_vllm_rl/weights/converter.py
Comment (tianyu-l) [inline torchtitan/experiments/deterministic_vllm_rl/weights/converter.py:1]: I wonder if we could just reuse https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/qwen3/model/state_dict_adapter.py

If `to_hf` / `from_hf` are not exactly what vllm needs, we should have `to_vllm` and `from_vllm`. cc @wwwjn

File: torchtitan/experiments/deterministic_vllm_rl/simple_rl.py [hunk 14]
```diff
@@ -0,0 +1,850 @@
...
+
+        # Save to temp model directory
+        checkpoint_path = os.path.join(self.temp_model_dir, "model.safetensors")
+
+        # Update the shard files that vLLM will actually load
+        # We need to split our weights to match the original 2-shard structure
+        import glob
+        import json
+
+        shard_files = sorted(glob.glob(os.path.join(self.temp_model_dir, "model-*.safetensors")))
```
Comment (tianyu-l) [inline torchtitan/experiments/deterministic_vllm_rl/simple_rl.py]: oh this weight sync is
- publishing weights to temp checkpoint files
- init vLLM from those files

I'd imagine there are better ways to do it. cc @LucasLLC

File: torchtitan/experiments/deterministic_vllm_rl/simple_rl.py [hunk 15]
```diff
@@ -0,0 +1,850 @@
...
+
+        This converts weights to vLLM format, saves them, and reloads using
+        vLLM's reload_weights() API after updating the model path config.
+
+        Args:
+            vllm_compat_state: vLLM-compat model state dict (with gate_up_proj/down_proj)
+        """
+        # Convert vLLM-compat -> TorchTitan -> vLLM
+        titan_state = vllm_compat_to_torchtitan(vllm_compat_state)
+        vllm_state = torchtitan_to_vllm(titan_state)
```
Comment (tianyu-l) [inline torchtitan/experiments/deterministic_vllm_rl/simple_rl.py]: this is confusing lol. Ideally we could make this minimal, by running torchtitan model on vLLM.

File: torchtitan/experiments/deterministic_vllm_rl/models/attention.py (hunk 6 above)
Comment (tianyu-l) [reply torchtitan/experiments/deterministic_vllm_rl/models/attention.py:60]: I still have this question, lol

File: torchtitan/experiments/deterministic_vllm_rl/batch_invariant_backward.py (hunk 3 above)
Comment (tianyu-l) [reply torchtitan/experiments/deterministic_vllm_rl/batch_invariant_backward.py]: can we remove this function?

File: torchtitan/experiments/deterministic_vllm_rl/simple_rl.py (hunk 15 above)
Comment (bwasti) [reply torchtitan/experiments/deterministic_vllm_rl/simple_rl.py]: vllm is far too optimized for this to be useful in a general context.  we don't get free speedups from using vllm (its not a compiler), we get speedups from vllm building and maintaining fast versions of models.

it can be demonstrated ([here](https://github.com/bwasti/vllm/commit/f6a598b730b3118dd7d65c4bdbb51cd45500133d#diff-b0d62052f3a5a808bc60a3c8380b1d5decbf2d31884687b95a6d94922cd889e7)) but it's very messy to do and I don't think it will scale with new features

File: torchtitan/experiments/deterministic_vllm_rl/models/attention.py (hunk 6 above)
Comment (bwasti) [reply torchtitan/experiments/deterministic_vllm_rl/models/attention.py:60]: ah, the FlashAttn3Function needs some special binary (that has a backward exposed) which is actually *not* shipped in nightly. the FlashAttnWithBackward was needed and I forgot to delete FlashAttn3Function

File: torchtitan/experiments/deterministic_vllm_rl/batch_invariant_backward.py (hunk 12 above)
Comment (bwasti) [reply torchtitan/experiments/deterministic_vllm_rl/batch_invariant_backward.py]: removed

File: torchtitan/experiments/deterministic_vllm_rl/simple_rl.py (hunk 14 above)
Comment (bwasti) [reply torchtitan/experiments/deterministic_vllm_rl/simple_rl.py]: the builtin reload is used @teja-rao swapped it

File: torchtitan/experiments/deterministic_vllm_rl/models/attention.py (hunk 6 above)
Comment (fegin) [reply torchtitan/experiments/deterministic_vllm_rl/models/attention.py:60]: Maybe add a TODO to convert this op to a custom op, which has a better composability with torch.compile. https://docs.pytorch.org/tutorials/advanced/python_custom_ops.html#python-custom-ops-tutorial

File: torchtitan/experiments/deterministic_vllm_rl/models/qwen3/model_vllm_compat.py [hunk 16]
```diff
@@ -0,0 +1,368 @@
...
+            self.model_args.max_seq_len,
+            self.model_args.rope_theta,
+        )
+
+    def get_attention_masks(
+        self,
+        input_batch: torch.Tensor,
+        tokenizer: BaseTokenizer,
+        extra_inputs: dict[str, torch.Tensor] | None = None,
+    ) -> AttentionMasksType:
```
Comment (fegin) [inline torchtitan/experiments/deterministic_vllm_rl/models/qwen3/model_vllm_compat.py]:
```suggestion
    ) -> AttentionMasksType | None:
```

**Comment by bwasti** [comment]
added TODOs to the readme
